src/lib/provable/test/sha2.unit-test.ts

```diff
@@ -0,0 +1,309 @@
+      },
+    ];
+  }
+}
+
+// SHA2-256 TESTS
+{
+  sample(Random.nat(400), 5).forEach((preimageLength) => {
+    let inputBytes = bytes(preimageLength);
+    let outputBytes = bytes(256 / 8);
+
+    equivalentProvable({ from: [inputBytes], to: outputBytes, verbose: true })(
+      (x) => nobleSha256(x),
+      (x) => Gadgets.SHA2.hash(256, x),
+      `sha256 preimage length ${preimageLength}`
+    );
+  });
+
+  const Sha2_256Program = ZkProgram({
+    name: `sha2-256`,
+    publicOutput: Bytes(32),
+    methods: {
+      sha256: {
+        privateInputs: [Bytes(192)],
+        async method(preImage: Bytes) {
+          return {
+            publicOutput: Gadgets.SHA2.hash(256, preImage),
+          };
+        },
+      },
+    },
+  });
+
+  const RUNS = 2;
+
+  await Sha2_256Program.compile();
+
+  await equivalentAsync(
+    {
+      from: [bytes(192)],
+      to: bytes(32),
+    },
+    { runs: RUNS }
+  )(nobleSha256, async (x) => {
+    const { proof } = await Sha2_256Program.sha256(x);
+    await Sha2_256Program.verify(proof);
+    return proof.publicOutput;
+  });
+
+  for (let { preimage, hash } of testVectors()) {
+    let actual = Gadgets.SHA2.hash(256, Bytes.fromString(preimage));
+    expect(actual.toHex()).toEqual(hash);
+  }
+
+  function testVectors() {
+    return [
+      {
+        preimage: 'abc',
+        hash: 'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad',
+      },
+      {
+        preimage: 'a'.repeat(1000000),
+        hash: 'cdc76e5c9914fb9281a1c7e284d73e67f1809a48a497200e046d39ccc7112cd0',
+      },
+      {
+        preimage: '',
+        hash: 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
+      },
+      {
+        preimage:
+          'abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmnoijklmnopjklmnopqklmnopqrlmnopqrsmnopqrstnopqrstu',
+        hash: 'cf5b16a778af8380036ce59e7b0492370b249b11e8f07a51afac45037afee9d1',
+      },
+    ];
+  }
+}
+
+// SHA2-384 TESTS
+{
+  sample(Random.nat(400), 5).forEach((preimageLength) => {
+    let inputBytes = bytes(preimageLength);
+    let outputBytes = bytes(384 / 8);
+
+    equivalentProvable({ from: [inputBytes], to: outputBytes, verbose: true })(
+      (x) => nobleSha384(x),
+      (x) => Gadgets.SHA2.hash(384, x),
+      `sha384 preimage length ${preimageLength}`
+    );
+  });
+
+  const Sha2_384Program = ZkProgram({
+    name: `sha2-384`,
+    publicOutput: Bytes(48),
+    methods: {
+      sha384: {
+        privateInputs: [Bytes(192)],
+        async method(preImage: Bytes) {
+          return {
+            publicOutput: Gadgets.SHA2.hash(384, preImage),
+          };
+        },
+      },
+    },
+  });
+
+  const RUNS = 2;
+
+  await Sha2_384Program.compile();
+
+  await equivalentAsync(
+    {
+      from: [bytes(192)],
+      to: bytes(48),
+    },
+    { runs: RUNS }
+  )(nobleSha384, async (x) => {
+    const { proof } = await Sha2_384Program.sha384(x);
+    await Sha2_384Program.verify(proof);
```

Comment:
Nit: is there a convention to check it also verifies? I added on my side.
```
const is_verified = await Sha2_384Program.verify(proof);
expect(is_verified);
```

I don't know if there is another place where we actually test the fact it verifies.

Reply:
Good point.. we are not doing that currently, anyway, but we probably should start doing that for completness altought I am not completely sure if Pickles or Kimchi checks if a proof is valid after generating it as a sanity check already